File: clients/android/app/src/main/kotlin/com/lextures/android/features/boards/BoardSurface.kt
package com.lextures.android.features.boards

import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.lextures.android.core.lms.ArrangeBoardPostBody
import com.lextures.android.core.lms.Board
import com.lextures.android.core.lms.BoardLayout
import com.lextures.android.core.lms.BoardPost
import com.lextures.android.core.lms.BoardReactionMode
import com.lextures.android.core.lms.BoardSection
import com.lextures.android.core.lms.BoardSortMode
import com.lextures.android.core.lms.BoardsLogic
import com.lextures.android.features.boards.layouts.CanvasLayout
import com.lextures.android.features.boards.layouts.ColumnsLayout
import com.lextures.android.features.boards.layouts.GridLayout
import com.lextures.android.features.boards.layouts.MapLayout
import com.lextures.android.features.boards.layouts.StreamLayout
import com.lextures.android.features.boards.layouts.TimelineLayout
import com.lextures.android.features.boards.layouts.WallLayout

/** Dispatches to a mobile layout renderer based on `board.layout` (VC.M3). */
@Composable
fun BoardSurface(
    board: Board,
    posts: List<BoardPost>,
    sections: List<BoardSection>,
    sortMode: BoardSortMode = BoardSortMode.Newest,
    canManage: Boolean,
    currentUserId: String?,
    onEdit: (BoardPost) -> Unit,
    onDelete: (BoardPost) -> Unit,
    onArrange: (BoardPost, ArrangeBoardPostBody) -> Unit,
    onCreateSection: ((String) -> Unit)? = null,
    onDeleteSection: ((String) -> Unit)? = null,
    modifier: Modifier = Modifier,
) {
    val reactionMode = BoardReactionMode.fromApi(board.reactionMode)
    when (BoardsLogic.resolveLayout(board.layout)) {
        BoardLayout.Wall -> WallLayout(
            posts = BoardsLogic.sortedPosts(posts, sortMode, reactionMode),
            sections = sections,
            board = board,
            canManage = canManage,
            currentUserId = currentUserId,
            onEdit = onEdit,
            onDelete = onDelete,
            onArrange = onArrange,
            modifier = modifier,
        )
        BoardLayout.Stream -> StreamLayout(
            posts = BoardsLogic.sortedPosts(posts, sortMode, reactionMode),
            sections = sections,
            board = board,
            canManage = canManage,
            currentUserId = currentUserId,
            onEdit = onEdit,
            onDelete = onDelete,
            onArrange = onArrange,
            modifier = modifier,
        )
        BoardLayout.Grid -> GridLayout(
            posts = BoardsLogic.sortedPosts(posts, sortMode, reactionMode),
            sections = sections,
            board = board,
            canManage = canManage,
            currentUserId = currentUserId,
            onEdit = onEdit,
            onDelete = onDelete,
            onArrange = onArrange,
            modifier = modifier,
        )
        BoardLayout.Columns -> ColumnsLayout(
            posts = posts,
            sections = sections,
            board = board,
            canManage = canManage,
            currentUserId = currentUserId,
            onEdit = onEdit,
            onDelete = onDelete,
            onArrange = onArrange,
            onCreateSection = onCreateSection,
            onDeleteSection = onDeleteSection,
            modifier = modifier,
        )
        BoardLayout.Canvas -> CanvasLayout(
            posts = posts,
            sections = sections,
            board = board,
            canManage = canManage,
            currentUserId = currentUserId,
            onEdit = onEdit,
            onDelete = onDelete,
            onArrange = onArrange,
            modifier = modifier,
        )
        BoardLayout.Timeline -> TimelineLayout(
            posts = posts,
            sections = sections,
            board = board,
            canManage = canManage,
            currentUserId = currentUserId,
            onEdit = onEdit,
            onDelete = onDelete,
            onArrange = onArrange,
            modifier = modifier,
        )
        BoardLayout.Map -> MapLayout(
            posts = posts,
            sections = sections,
            board = board,
            canManage = canManage,
            currentUserId = currentUserId,
            onEdit = onEdit,
            onDelete = onDelete,
            onArrange = onArrange,
            modifier = modifier,
        )
    }
}

@Composable
fun BoardPostCardSlot(
    post: BoardPost,
    siblings: List<BoardPost>,
    sections: List<BoardSection>,
    board: Board,
    canManage: Boolean,
    currentUserId: String?,
    onEdit: (BoardPost) -> Unit,
    onDelete: (BoardPost) -> Unit,
    onArrange: (BoardPost, ArrangeBoardPostBody) -> Unit,
    showTimeline: Boolean = false,
    showMap: Boolean = false,
    modifier: Modifier = Modifier,
) {
    BoardPostCard(
        post = post,
        canEdit = BoardsLogic.canEditOrDeletePost(post, currentUserId, canManage),
        onEdit = { onEdit(post) },
        onDelete = { onDelete(post) },
        modifier = modifier,
        canArrange = BoardsLogic.canArrangePost(post, board, currentUserId, canManage),
        canManageBoard = canManage,
        currentUserId = currentUserId,
        reactionMode = BoardReactionMode.fromApi(board.reactionMode),
        canInteract = BoardsLogic.canInteract(board) &&
            BoardsLogic.canWriteInteractions(board, canManage),
        assignmentLinked = BoardsLogic.assignmentLinked(board),
        sections = sections,
        siblings = siblings,
        showTimelineArrange = showTimeline,
        showMapArrange = showMap,
        onArrange = { onArrange(post, it) },
    )
}

File: clients/ios/Lextures/Features/Boards/BoardSurface.swift
import SwiftUI

/// Dispatches to a mobile layout renderer based on `board.layout` (VC.M3).
struct BoardSurface: View {
    let board: Board
    let posts: [BoardPost]
    let sections: [BoardSection]
    var sortMode: BoardSortMode = .newest
    var canManage: Bool = false
    var currentUserId: String?
    var onEdit: (BoardPost) -> Void
    var onDelete: (BoardPost) -> Void
    var onArrange: (BoardPost, ArrangeBoardPostInput) -> Void
    var onCreateSection: ((String) -> Void)?
    var onDeleteSection: ((String) -> Void)?

    private var layout: BoardLayout {
        BoardsLogic.resolveLayout(board.layout)
    }

    private var reactionMode: BoardReactionMode {
        BoardReactionMode.fromAPI(board.reactionMode)
    }

    var body: some View {
        Group {
            switch layout {
            case .wall:
                WallLayoutView(
                    posts: BoardsLogic.sortedPosts(posts, mode: sortMode, reactionMode: reactionMode),
                    sections: sections,
                    board: board,
                    canManage: canManage,
                    currentUserId: currentUserId,
                    onEdit: onEdit,
                    onDelete: onDelete,
                    onArrange: onArrange
                )
            case .stream:
                StreamLayoutView(
                    posts: BoardsLogic.sortedPosts(posts, mode: sortMode, reactionMode: reactionMode),
                    sections: sections,
                    board: board,
                    canManage: canManage,
                    currentUserId: currentUserId,
                    onEdit: onEdit,
                    onDelete: onDelete,
                    onArrange: onArrange
                )
            case .grid:
                GridLayoutView(
                    posts: BoardsLogic.sortedPosts(posts, mode: sortMode, reactionMode: reactionMode),
                    sections: sections,
                    board: board,
                    canManage: canManage,
                    currentUserId: currentUserId,
                    onEdit: onEdit,
                    onDelete: onDelete,
                    onArrange: onArrange
                )
            case .columns:
                ColumnsLayoutView(
                    posts: posts,
                    sections: sections,
                    board: board,
                    canManage: canManage,
                    currentUserId: currentUserId,
                    onEdit: onEdit,
                    onDelete: onDelete,
                    onArrange: onArrange,
                    onCreateSection: onCreateSection,
                    onDeleteSection: onDeleteSection
                )
            case .canvas:
                CanvasLayoutView(
                    posts: posts,
                    sections: sections,
                    board: board,
                    canManage: canManage,
                    currentUserId: currentUserId,
                    onEdit: onEdit,
                    onDelete: onDelete,
                    onArrange: onArrange
                )
            case .timeline:
                TimelineLayoutView(
                    posts: posts,
                    sections: sections,
                    board: board,
                    canManage: canManage,
                    currentUserId: currentUserId,
                    onEdit: onEdit,
                    onDelete: onDelete,
                    onArrange: onArrange
                )
            case .map:
                MapLayoutView(
                    posts: posts,
                    sections: sections,
                    board: board,
                    canManage: canManage,
                    currentUserId: currentUserId,
                    onEdit: onEdit,
                    onDelete: onDelete,
                    onArrange: onArrange
                )
            }
        }
        .accessibilityLabel(L.text("mobile.boards.layout.\(layout.rawValue)"))
    }
}

struct BoardPostCardSlot: View {
    let post: BoardPost
    let siblings: [BoardPost]
    let sections: [BoardSection]
    let board: Board
    var canManage: Bool
    var currentUserId: String?
    var showTimeline: Bool = false
    var showMap: Bool = false
    var onEdit: (BoardPost) -> Void
    var onDelete: (BoardPost) -> Void
    var onArrange: (BoardPost, ArrangeBoardPostInput) -> Void

    private var canEdit: Bool {
        BoardsLogic.canEditOrDeletePost(post: post, currentUserId: currentUserId, canManage: canManage)
    }

    private var canArrange: Bool {
        BoardsLogic.canArrangePost(
            post: post,
            board: board,
            currentUserId: currentUserId,
            canManage: canManage
        )
    }

    var body: some View {
        BoardPostCard(
            post: post,
            canEdit: canEdit,
            canArrange: canArrange,
            canManageBoard: canManage,
            currentUserId: currentUserId,
            reactionMode: BoardReactionMode.fromAPI(board.reactionMode),
            canInteract: BoardsLogic.canInteract(board: board)
                && BoardsLogic.canWriteInteractions(board: board, canManage: canManage),
            assignmentLinked: BoardsLogic.assignmentLinked(board),
            sections: sections,
            siblings: siblings,
            showTimelineArrange: showTimeline,
            showMapArrange: showMap,
            onEdit: { onEdit(post) },
            onDelete: { onDelete(post) },
            onArrange: { onArrange(post, $0) }
        )
    }
}
